indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamEndSequenceNumbers.java

```diff
@@ -147,6 +148,40 @@ public SeekableStreamSequenceNumbers<PartitionIdType, SequenceOffsetType> plus(
     }
   }
 
+  @Override
+  public int compareTo(SeekableStreamSequenceNumbers<PartitionIdType, SequenceOffsetType> other)
+  {
+    if (this.getClass() != other.getClass()) {
+      throw new IAE(
+          "Expected instance of %s, got %s",
+          this.getClass().getName(),
+          other.getClass().getName()
+      );
+    }
+
+    final SeekableStreamEndSequenceNumbers<PartitionIdType, SequenceOffsetType> otherStart =
+        (SeekableStreamEndSequenceNumbers<PartitionIdType, SequenceOffsetType>) other;
+
+    if (stream.equals(otherStart.stream)) {
+      //Same stream, compare the offset
+      AtomicReference<Boolean> res = new AtomicReference<>(false);
+      partitionSequenceNumberMap.forEach(
+          (partitionId, sequenceOffset) -> {
+            if (otherStart.partitionSequenceNumberMap.get(partitionId) != null && Long.parseLong(String.valueOf(sequenceOffset)) > Long.parseLong(String.valueOf(otherStart.partitionSequenceNumberMap.get(partitionId)))) {
```

Comment:
I'm not sure we can assume sequenceOffsets can always be parsed as long. This seems to be a fairly broad assumption. Someone more familiar with the Kinesis supervisor should chime in.

Reply:
This could be done. Two methods should be added to DataSourceMetadata
```
isComparable (default return false)
compare() (default return 0)
```

These two methods will be implemented only by the KafkaDataSourceMetadata. In that class, you can parse the offset to long. 

In the IndexerSQLMetadataStorageCoordinator, you would first check isComparable before calling compare method. 

How does this sound?

Reply:
or you can have just KafkaDataSourceMetadata implement Comparable. Then you would check if the metadata is of type comparable and call compare method. Then you don't need `isComparable` method and the interface doesn't change either. That's actually better and easy to change in the future.

Reply:
@abhishekagarwal87, I have implemented comparable method only in kafkaDataSourceMetadata. And kinesisDataSourceMetadata will return 0(default).  PTAL

Reply:
how would we handle this potential failure scenario in Kinesis then? It seems that the issue is not specific to Kafka and could happen there as well.

Reply:
It could, yes. The solution here is to either retry no matter how current and new offsets compare. Or we could add the comparison to Kinesis too - https://docs.aws.amazon.com/kinesis/latest/APIReference/API_SequenceNumberRange.html - The sequence numbers seem to be numbers in string form but can contain up to 128 digits. It should be possible to compare them without converting them to a number.

Reply:
The class `OrderedSequenceNumber` may already have the implementation to compare Kafka and Kinesis sequence numbers.

Reply:
No, OrderedSequenceNumber does not have implementation to compare.

Reply:
@Pankaj260100, Don't `KafkaSequenceNumber` and `KinesisSequenceNumber` which extend this class handle the comparison at an individual sequence number's level?